### PR TITLE
[5.1] Allow custom fields form be manipulates like category form

### PR DESCRIPTION
**Pull request by bembelimen**
### Summary of Changes
Allow a component to manipulate the custom field form like you [can do with category view](https://github.com/joomla/joomla-cms/blob/5.1-dev/administrator/components/com_categories/src/Model/CategoryModel.php#L431-L460)

### Testing Instructions
- Apply patch
- Open a field form from com_content in the backend
- Check that you get the form [here](https://github.com/joomla/joomla-cms/blob/5.1-dev/administrator/components/com_content/src/Extension/ContentComponent.php#L371) and can manipulate it.

```$form->getName()``` should give ```com_fields.field.com_content.article```

```$form->removeField('display', 'params');``` removes the "Automatic Display" field

### Actual result BEFORE applying this Pull Request
No way to manipulate the form via PHP directly

### Expected result AFTER applying this Pull Request
You can manipulate the form

### CODE REVIEW

File: libraries/src/Fields/FieldsServiceInterface.php [hunk 1]
```diff
@@ -41,4 +43,16 @@ public function validateSection($section, $item = null);
...
      */
     public function getContexts(): array;
+
+    /**
+     * Prepares the category form
+     *
+     * @param   Form          $form  The form to change
+     * @param   array|object  $data  The form data
+     *
+     * @return   void
```
Comment (brianteeman) [inline libraries/src/Fields/FieldsServiceInterface.php]:
```suggestion
     * @return  void
```

**Comment by brianteeman** [comment]
would love to test this if you could provide some realistic test instructions

File: libraries/src/Fields/FieldsServiceInterface.php (hunk 1 above)
Comment (bembelimen) [reply libraries/src/Fields/FieldsServiceInterface.php]: Thx

**Comment by bembelimen** [comment]
> would love to test this if you could provide some realistic test instructions

Done

**Comment by brianteeman** [comment]
> > would love to test this if you could provide some realistic test instructions
> 
> Done

I will have to pass on testing this then

File: administrator/components/com_fields/src/Model/FieldModel.php [hunk 2]
```diff
@@ -1108,6 +1109,37 @@ function () use ($component, $section) {
...
+
+                    // Check for an error.
+                    if ($form instanceof \Exception) {
+                        $this->setError($form->getMessage());
+
+                        return false;
+                    }
+                }
+            }
+        }
```
Comment (HLeithner) [inline administrator/components/com_fields/src/Model/FieldModel.php]:
```suggestion
        }
```

no reason to support non-bootable components anymore for new features.

Comment (bembelimen) [reply administrator/components/com_fields/src/Model/FieldModel.php]: True!

File: administrator/components/com_fields/src/Model/FieldModel.php [hunk 3]
```diff
@@ -1108,6 +1109,12 @@ function () use ($component, $section) {
             }
         }
 
+        $componentInterface = Factory::getApplication()->bootComponent($component);
+
+        if ($componentInterface instanceof FieldsServiceInterface) {
+            $componentInterface->prepareForm($form, $data);
+        }
```
Comment (HLeithner) [inline administrator/components/com_fields/src/Model/FieldModel.php]:
```suggestion
        $componentBooted = Factory::getApplication()->bootComponent($component);

        if ($componentBooted instanceof FieldsServiceInterface) {
            $componentBooted->prepareForm($form, $data);
        }
```

You don't have an interface in the variable so calling it ...Interface is missleading.

**Comment by chmst** [comment]
I have tested this item :white_check_mark: successfully on 343863bf4783176a95bb7ae4266864a9a694d8da<br /><br />Tested following the instruction and was abe to manipuate coustoms fields. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42510">issues.joomla.org/tracker/joomla-cms/42510</a>.</sub>

**Comment by HLeithner** [comment]
I have tested this item :white_check_mark: successfully on a86650fd304c01634065b531f36e1c242e7336a9<br /><br />Tested with com_contact, call works side didn't break<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42510">issues.joomla.org/tracker/joomla-cms/42510</a>.</sub>

**Comment by richard67** [comment]
I've restored @chmst 's test result in the issue tracker since the commit which has invalidated the test counter was just a clean branch update.

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42510">issues.joomla.org/tracker/joomla-cms/42510</a>.</sub>

**Comment by LadySolveig** [comment]
Thank you! And also for support and testing @chmst @HLeithner @richard67

**Pull request merged**